File: blackjack/elixir/outcomes2.ex

defmodule Outcomes do
  def partitions(cards, subtotal) do
    Enum.reduce([0,1,2,3,4,5,6,7,8,9], 0, fn(i, acc) ->
      case elem(cards,i)>0 do
	:true ->
	  cond do
	    subtotal+i+1 > 21 -> acc
	    subtotal+i+1==21 -> acc+1
	    subtotal+i+1 < 21 ->
	      acc+1+partitions(put_elem(cards,i,elem(cards,i)-1),subtotal+i+1)
	  end
	:false -> acc
      end
    end)
  end
end

deck = Tuple.append(4 |> Tuple.duplicate(9), 16)

d =
  Enum.sum(
    for i <- 0..9 do
      new_deck = put_elem(deck, i, elem(deck, i)-1)
      p =
      Enum.sum(
	for j <- 0..9 do
	  Outcomes.partitions(put_elem(new_deck, j, elem(new_deck, j)-1), j+1)
	end)
      IO.puts "Dealer showing #{i} partitions = #{p}"
      p
    end
  )

IO.puts "Total partitions = #{d}"

File: blackjack/python/outcomes2.py
#!/usr/bin/env python

def partitions(cards, subtotal):
    m = 0
    # Hit
    for i in xrange(10):
        if (cards[i]>0):
            total = subtotal+i+1
            if (total < 21):
                # Stand
                m += 1
                # Hit again
                cards[i] -= 1
                m += partitions(cards, total)
                cards[i] += 1
            elif (total==21):
                # Stand; hit again is an automatic bust
                m += 1
                break
    return(m)

deck = ([4]*9)
deck.append(16)
d = 0

for i in xrange(10):

    # Dealer showing

    deck[i] -= 1

    p = 0
    for j in xrange(10):
        deck[j] -= 1
        p += partitions(deck, j+1)
        deck[j] += 1

    print "Dealer showing ",i," partitions =",p
    d += p

    deck[i] += 1

print "Total partitions =",d
